### PR TITLE
add new, optimized tx queue

### PR DESCRIPTION
**Pull request by jack60612**
- [x] Working new queue
- [x] Priority for local tx's 
- [x] Unit Tests

### CODE REVIEW

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:48]: did you consider making `put()` sleep here if the queue is full? That might simplify the API, you may no longer need `empty()`, `full()` or `qsize()`

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:54]: I think the item is removed here too, right? I think `pop()` would be a better name, but maybe if the python convention is `get()` we should go with that.

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:71]: this doesn't look right. is this really necessary?

**Review by arvidn — CHANGES_REQUESTED** [review]
You need unit tests too

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:17]: both async and normal queue have built in syncing

Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:17]: good point about the peer id

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:19]: i had a limit on both but ill get rid of the main limit

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:48]: yeah i just changed it up and simplified the api

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:54]: ill change it to pop

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:71]: don't we need to delete old queues?

Comment (arvidn) [reply chia/full_node/tx_processing_queue.py:69]: I thought your idea of removing empty queues once per loop was good. It strikes a balance between capping resource use and still give peers some time to reuse the previous one if there's another incoming transaction. I also think it is a simpler design to not have yet another task running

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:73]: this ends up being a linear search still, which isn't great. On the other hand, the size of this list will probably rarely be very long anyway, so if you can't think of a simple fix, it's probably fine to leave it like this

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:75]: this will bias the queues we pick from to favor lower indices. I don't think this is OK. Rather than doing this, make sure to update the `_list_iterator` when removing an item with a lower or equal index.

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:21]: I think the term "iterator" refers to specific kinds of objects in python (and many other languages too). I think "cursor" would be a better name. maybe even have a comment to describe what it's used for.

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:25]: I think this should be a `queue.SimpleQueue`. We don't use any of the synchronization that comes with `asyncio.Queue`

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [inline chia/full_node/tx_processing_queue.py:62]: I think this would throw an `IndexError` exception. but you'll discover that when you add the tests.

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:69]: ok ill revert back to that.

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:46]: before these changes, if the queue was full we didnt add it, i thought this was a good idea, unless you think we dont want that.

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:73]: yeah there should always be under 100 peers so it should be fine.

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:25]: SimpleQueue Doesn't have a peer transaction limit

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:62]: it will, oversight on my part.

Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:62]: > I think this would throw an `IndexError` exception. but you'll discover that when you add the tests.

i don't think it will, however it is broken, so im fixing it.

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:21]: yeah your right.

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [reply chia/full_node/tx_processing_queue.py:46]: actually, in an ideal world, we would apply back-pressure to the peer and reject transactions once the queue is full. Maybe that's better. In that case, none of the queues should be `asyncio.Queue` but just `queue.SimpleQueue`

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [reply chia/full_node/tx_processing_queue.py:62]: you increment `_list_iterator` without checking if you just past the last item of the list, then you access the list at that index.

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [reply chia/full_node/tx_processing_queue.py:25]: that's fine. it has a length

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [reply chia/full_node/tx_processing_queue.py:73]: I would assume closer to 1000 for a big node. Anyway, one simple change to turn this from O(n^2) to O(n) would be to not update `_index_to_peer_map` as you go, but rebuild it from scratch when you're done.

Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:73]: > I would assume closer to 1000 for a big node. Anyway, one simple change to turn this from O(n^2) to O(n) would be to not update `_index_to_peer_map` as you go, but rebuild it from scratch when you're done.

i did realize that i was doing remove, so i changed the loop a bit to loop off of the list instead of the dict.

Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:73]: @arvidn after looking here: [https://wiki.python.org/moin/TimeComplexity](https://wiki.python.org/moin/TimeComplexity) I don't think I have found a better way to do it.

Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:73]: @arvidn after looking here: [https://wiki.python.org/moin/TimeComplexity](https://wiki.python.org/moin/TimeComplexity) I don't think I have found a better way to do it.

**Comment by jack60612** [comment]
im going to finish this and add unit tests later, ill let you know when i want another review @arvidn

File: chia/full_node/tx_processing_queue.py
Comment (arvidn) [reply chia/full_node/tx_processing_queue.py:73]: iterating over the list would have better complexity. this is untested, but something like this:

```
new_peer_map = []
for idx, peer_id in enumerate(self._index_to_peer_map):
    if self._queue_dict[peer_id].empty():
        self._queue_dict.pop(peer_id)
        if idx < self._list_iterator:
            self._list_iterator -= 1
    else:
        new_peer_map.append(peer_id)
self._index_to_peer_map = new_peer_map
```

Comment (jack60612) [reply chia/full_node/tx_processing_queue.py]: > iterating over the list would have better complexity. this is untested, but something like this:
> 
> ```
> new_peer_map = []
> for idx, peer_id in enumerate(self._index_to_peer_map):
>     if self._queue_dict[peer_id].empty():
>         self._queue_dict.pop(peer_id)
>         if idx < self._list_iterator:
>             self._list_iterator -= 1
>     else:
>         new_peer_map.append(peer_id)
> self._index_to_peer_map = new_peer_map
> ```

i kept the list building part, but dropped the enumerate part, as we are resetting to zero anyway.

Comment (arvidn) [reply chia/full_node/tx_processing_queue.py]: right

**Comment by jack60612** [comment]
> im going to finish this and add unit tests later, ill let you know when i want another review @arvidn

rdy for review

**Review by arvidn — COMMENTED** [review]
I think the tests could be simpler (and a bit easier to follow). I believe there are a few cases missing:

* test when a peer id's queue is full. expect some kind of feedback (probably an exception)
* test that one peer posting 2 items and a different peer posts 1 item, still are serviced fairly
* it might make sense to ensure the `pop()` function does in fact sleep when the queue is empty, and wakes up when an item is pushed.

File: tests/core/full_node/test_tx_processing_queue.py
Comment (jack60612) [reply tests/core/full_node/test_tx_processing_queue.py:109]: > I think you should remove this class and make all the functions free functions. that way you'll save one level of indentation too

we always use classes when using tests, so i think we should keep it.

File: chia/full_node/tx_processing_queue.py
Comment (jack60612) [reply chia/full_node/tx_processing_queue.py:50]: I'm going to add this and then add an error to the message queue for the local wallet, but the current blockchain protocol doesn't have responses for these types of messages and our node simply dropped the tx's if they were full.

**Comment by jack60612** [comment]
> I think the tests could be simpler (and a bit easier to follow). I believe there are a few cases missing:
> 
> * test when a peer id's queue is full. expect some kind of feedback (probably an exception)
> * test that one peer posting 2 items and a different peer posts 1 item, still are serviced fairly
> * it might make sense to ensure the `pop()` function does in fact sleep when the queue is empty, and wakes up when an item is pushed.

test 1 and 3 added, and existing were cleaned up.

File: chia/full_node/full_node_api.py
Comment (arvidn) [inline chia/full_node/full_node_api.py:256]: right, I see. peers aren't sending these TXs to us, we're requesting them. In the future we might want to stop requesting more TXs once we get close to the limit

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [reply tests/core/full_node/test_tx_processing_queue.py:61]: it's used in many places, most likely because of some original misunderstandings followed by copy+paste. We shouldn't keep bad habits just because they're habits.

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [inline tests/core/full_node/test_tx_processing_queue.py:58]: I think all of the code above is a distraction, diluting these tests and making them harder to follow. the `TransactionQueueEntry` does not need to contain a peer_id or a `SpendBundle`. I think you should create a dummy class that just has some identifier in it (to tell them apart), then cast that to a `TransactionQueueEntry`. It should work at run-time because the queue shouldn't be looking at any of its fields.

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [inline tests/core/full_node/test_tx_processing_queue.py:117]: is this actually waiting in wall-clock time? If so, I think we should test this without sleeping. e.g. we could create a task and ensure it's blocked in the `pop()`

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [inline tests/core/full_node/test_tx_processing_queue.py:111]: it seems to me that the first part of this test, pushing 100 items and then popping them is redundant. You already have a test that does that

File: tests/core/full_node/test_tx_processing_queue.py
Comment (jack60612) [reply tests/core/full_node/test_tx_processing_queue.py:61]: > > I think you should remove this class and make all the functions free functions. that way you'll save one level of indentation too
> 
> we always use classes when using tests, so i think we should keep it.

unless there is a specific reason not to?

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [inline tests/core/full_node/test_tx_processing_queue.py:67]: I don't see a test that ensures local TXs take precedence over TXs we receive from peers

**Review by arvidn — COMMENTED** [review]
generally it's not great practice to have tests verify internal state (and implementation details). Normally you want to just verify the observable behavior of the class

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [reply tests/core/full_node/test_tx_processing_queue.py:61]: in my mind, the main reasons not to are:
1. save one level of indentation
2. save lines of code
3. make it clear to the reader that the tests do not in fact share state (which could be a legitimate reason to use a class)

File: tests/core/full_node/test_tx_processing_queue.py
Comment (jack60612) [reply tests/core/full_node/test_tx_processing_queue.py:67]: i was about to add to add that lol, good catch.

File: tests/core/full_node/test_tx_processing_queue.py
Comment (jack60612) [reply tests/core/full_node/test_tx_processing_queue.py:61]: ok that makes sense, will do!

File: chia/full_node/full_node_api.py
Comment (jack60612) [reply chia/full_node/full_node_api.py:256]: i could add that

Comment (arvidn) [reply chia/full_node/full_node_api.py:256]: that should be a separate PR. we don't want to tie too many independent changes together

File: tests/core/full_node/test_tx_processing_queue.py
Comment (jack60612) [reply tests/core/full_node/test_tx_processing_queue.py:111]: we are testing the non high priority queue, all local / high priority txs bypass the round robin logic.

File: tests/core/full_node/test_tx_processing_queue.py
Comment (jack60612) [reply tests/core/full_node/test_tx_processing_queue.py:58]: changed that, thanks!

File: tests/core/full_node/test_tx_processing_queue.py
Comment (arvidn) [inline tests/core/full_node/test_tx_processing_queue.py:150]: this test could be as simple as:

* pushing 2 entries for peer A, one entry for B and 2 entries for C
* pop 3 items (ensuring it's from A, B and C)
* pop 2 items, ensuring it's from A and C (in that order)

It's problematic to have tests assert state of implementation details. One of the purposes of tests is to ensure the class behaves correctly after refactoring it, or changing the implementation strategy. With tests like this, one would have to re-write the tests as well, which defeats an important point of having them.

**Comment by arvidn** [comment]
This looks a lot better. I can't find anything obvious that's not coverered by a test. However, I think it might make sense to put a bit more work into making the tests simple (after all, we don't test the tests, we rely on them being simple enough to ensure their correctness by manual inspection).

Also, tests that access private members and test implementation details are fragile and not as useful as tests that ensure the observed behavior only.

**Review by arvidn — APPROVED** [review]

**Pull request merged**
